Sources/NetworkProtection/PacketTunnelProvider.swift

```diff
@@ -664,54 +669,70 @@ open class PacketTunnelProvider: NEPacketTunnelProvider {
 
     // MARK: - Tunnel Stop
 
-    open override func stopTunnel(with reason: NEProviderStopReason, completionHandler: @escaping () -> Void) {
+    @MainActor
+    open override func stopTunnel(with reason: NEProviderStopReason) async {
+        providerEvents.fire(.tunnelStopAttempt(.begin))
 
-        Task { @MainActor in
-            await stopMonitors()
+        os_log("Stopping tunnel with reason %{public}@", log: .networkProtection, type: .info, String(describing: reason))
 
-            connectionStatus = .disconnecting
-            os_log("Stopping tunnel with reason %{public}@", log: .networkProtection, type: .info, String(describing: reason))
+        do {
+            try await stopTunnel()
+            providerEvents.fire(.tunnelStopAttempt(.success))
+        } catch {
+            providerEvents.fire(.tunnelStopAttempt(.failure(error)))
+        }
 
-            adapter.stop { [weak self] error in
-                if let error {
-                    os_log("🔵 Failed to stop WireGuard adapter: %{public}@", log: .networkProtection, type: .info, error.localizedDescription)
-                    self?.debugEvents?.fire(error.networkProtectionError)
-                }
+        if case .superceded = reason {
+            self.notificationsPresenter.showSupersededNotification()
+        }
+    }
 
-                Task { [weak self] in
-                    if let self {
-                        self.handleAdapterStopped()
+    /// Do not cancel, directly... call this method so that the adapter and tester are stopped too.
+    @MainActor
+    private func cancelTunnel(with stopError: Error) async {
+        providerEvents.fire(.tunnelStopAttempt(.begin))
 
-                        if case .superceded = reason {
-                            self.notificationsPresenter.showSupersededNotification()
-                        }
-                    }
+        os_log("Stopping tunnel with error %{public}@", log: .networkProtection, type: .error, stopError.localizedDescription)
 
-                    completionHandler()
-                }
-            }
+        do {
+            try await stopTunnel()
+            providerEvents.fire(.tunnelStopAttempt(.success))
+        } catch {
+            providerEvents.fire(.tunnelStopAttempt(.failure(error)))
         }
+
+        cancelTunnelWithError(stopError)
     }
 
-    /// Do not cancel, directly... call this method so that the adapter and tester are stopped too.
+    // MARK: - Tunnel Stop: Support Methods
+
+    /// Do not call this directly.  Call `stopTunnel(with:)` or `cancelTunnel(with:)` instead.
+    ///
     @MainActor
-    private func cancelTunnel(with stopError: Error) async {
+    private func stopTunnel() async throws {
```

Comment:
I'd be tempted to rename this to clearly indicate it's not intended for direct use. Not a blocker though.